File: 021.hs




main :: IO ()
main = putStrLn $ show $ answer

answer :: Integer
answer = sum [x | x <- [2..10000] , isAmicable x]

isAmicable :: Integer -> Bool
isAmicable x = (x == (sum $ divisions $ sum $ divisions x)) && (not $ isPerfect x)

divisions :: Integer -> [Integer]
divisions n = [x | x <- [1..(n-1)] , n `mod` x == 0] 

isPerfect :: Integer -> Bool
isPerfect x = x == (sum $ divisions x)

File: 021.rb
# -*- coding: utf-8 -*-
require 'mathn'

class Array
  def sum
    inject(:+)
  end

  def product
    inject(:*)
  end

  def subsets
    (1..self.length).inject([]) { |r,l|
      r + combination(l).to_a
    }
  end
end

class Integer
  #約数を求める
  def divisions
    prime_divisions = prime_division.inject([]) {|r,x| r+[x.first]*x[1] }
    result = prime_divisions.subsets.map{|x| x.product }
    result << 1
    result.uniq.sort
  end

  #自身をのぞく約数を求める
  def divisions_without_self
    divisions.select {|x| x!=self }
  end

  #自身をのぞく約数の和を求める
  def sum_of_divisions
    divisions_without_self.sum
  end

  #友愛数かどうかの判定
  def amicable?
    self == self.sum_of_divisions.sum_of_divisions && ! perfect?
  end

  #完全数かどうかの判定
  def perfect?
    self == self.sum_of_divisions
  end
end

puts (2..10000).select { |x| x.amicable? }.sum
